File: src/main/java/com/itsci/mju/maebanjumpen/report/service/impl/ReportServiceImpl.java
package com.itsci.mju.maebanjumpen.report.service.impl;

import com.itsci.mju.maebanjumpen.hire.repository.HireRepository;
import com.itsci.mju.maebanjumpen.partyrole.repository.PartyRoleRepository;
import com.itsci.mju.maebanjumpen.penalty.repository.PenaltyRepository;
import com.itsci.mju.maebanjumpen.report.dto.ReportDTO;
import com.itsci.mju.maebanjumpen.exception.AlreadyReportedException;
import com.itsci.mju.maebanjumpen.mapper.ReportMapper;
import com.itsci.mju.maebanjumpen.entity.*;
import com.itsci.mju.maebanjumpen.report.repository.ReportRepository;
import com.itsci.mju.maebanjumpen.hire.service.impl.HireStatusUpdateService;
import com.itsci.mju.maebanjumpen.person.service.PersonService;
import com.itsci.mju.maebanjumpen.report.service.ReportService;
import lombok.RequiredArgsConstructor;
import org.hibernate.Hibernate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ReportServiceImpl implements ReportService {

    private final ReportRepository reportRepository;
    private final PartyRoleRepository partyRoleRepository;
    // ❌ ลบ private final HirerRepository hirerRepository;
    // ❌ ลบ private final HousekeeperRepository housekeeperRepository;
    private final PenaltyRepository penaltyRepository;
    private final HireRepository hireRepository;
    private final ReportMapper reportMapper;
    private final PersonService personService;

    // 🎯 Service สำหรับการจัดการการเปลี่ยนสถานะตามเวลา
    private final HireStatusUpdateService hireStatusUpdateService;

    /**
     * Initializes lazy-loaded collections/associations of the Report entity.
     * Hirer and Housekeeper are now initialized via the Hire object.
     */
    private void initializeReport(Report report) {
        if (report == null) return;
        Hibernate.initialize(report.getReporter());
        // ❌ ลบ: report.getHirer() และ report.getHousekeeper() ถูกลบออกจาก Report Entity แล้ว

        Hibernate.initialize(report.getPenalty());
        Hibernate.initialize(report.getHire());

        // 💡 ตรวจสอบและ Initialize Hirer/Housekeeper ผ่าน Hire object
        if (report.getHire() != null) {
            Hibernate.initialize(report.getHire().getHirer());
            Hibernate.initialize(report.getHire().getHousekeeper());

            // Initialize Person (หากต้องการดึงชื่อ/ข้อมูลพื้นฐาน)
            if (report.getHire().getHirer() != null && report.getHire().getHirer().getPerson() != null) {
                Hibernate.initialize(report.getHire().getHirer().getPerson());
            }
            if (report.getHire().getHousekeeper() != null && report.getHire().getHousekeeper().getPerson() != null) {
                Hibernate.initialize(report.getHire().getHousekeeper().getPerson());
            }
        }

        if (report.getReporter() != null && report.getReporter().getPerson() != null) {
            Hibernate.initialize(report.getReporter().getPerson());
        }
    }

    @Override
    public List<ReportDTO> getAllReports() {
        List<Report> reports = reportRepository.findAll();
        reports.forEach(this::initializeReport);
        return reportMapper.toDtoList(reports);
    }

    @Override
    public ReportDTO getReportById(int id) {
        Report report = reportRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Report not found with id: " + id));
        initializeReport(report);
        return reportMapper.toDto(report);
    }

    @Override
    @Transactional
    public ReportDTO createReport(ReportDTO reportDto) {
        if (reportDto.getHire() == null || reportDto.getHire().getHireId() == null) {
            throw new IllegalArgumentException("Hire object with ID is required.");
        }

        Integer reporterId = reportDto.getReporter() != null ? reportDto.getReporter().getId() : null;
        Integer hireId = reportDto.getHire().getHireId(); // ✅ ดึง Hire ID จาก HireDTO
        Integer penaltyId = reportDto.getPenalty() != null ? reportDto.getPenalty().getPenaltyId() : null;

        if (reporterId == null) {
            throw new IllegalArgumentException("Reporter ID is required in the reporter object.");
        }

        // 🛑 ตรวจสอบการรายงานซ้ำ
        if (reportRepository.findByHire_HireIdAndReporter_Id(hireId, reporterId).isPresent()) {
            throw new AlreadyReportedException("You have already submitted a report for this job.");
        }

        Report report = reportMapper.toEntity(reportDto);

        PartyRole existingReporter = partyRoleRepository.findById(reporterId)
                .orElseThrow(() -> new IllegalArgumentException("Reporter not found with ID: " + reporterId));

        Hire existingHire = hireRepository.findById(hireId)
                .orElseThrow(() -> new IllegalArgumentException("Hire not found with ID: " + hireId));

        // 2. ผูก Entity
        report.setReporter(existingReporter);
        report.setHire(existingHire);
        // ❌ ลบ: report.setHirer(...) และ report.setHousekeeper(...)
        report.setPenalty(penaltyId != null ? penaltyRepository.findById(penaltyId).orElse(null) : null);

        if (report.getReportStatus() == null || report.getReportStatus().isEmpty()) {
            report.setReportStatus("pending"); // สถานะของรายงาน
        }

        Report savedReport = reportRepository.save(report);

        // ----------------------------------------------------
        // ✅ LOGIC การเปลี่ยนสถานะงานจ้าง (อยู่ใน Transaction เดียวกัน)
        // ----------------------------------------------------

        // 3. เปลี่ยนสถานะงานจ้าง (Hire) เป็น "Reported" ทันที
        existingHire.setJobStatus("Reported");
        hireRepository.save(existingHire);

        System.out.println("LOG: Hire ID " + existingHire.getHireId() + " status set to 'Reported' and report saved.");


        // 4. กำหนดเวลาเปลี่ยนสถานะกลับเป็น "Completed" ใน 3 วินาที (Task Scheduler)
        if (existingHire.getHireId() != null) {
            hireStatusUpdateService.scheduleStatusRevert(
                    existingHire.getHireId(),
                    3L
            );
        }
        // ----------------------------------------------------

        initializeReport(savedReport);
        return reportMapper.toDto(savedReport);
    }


    @Override
    @Transactional
    public ReportDTO updateReport(int id, ReportDTO reportDto) {
        Report existingReport = reportRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Report not found with id: " + id));

        // 1. อัปเดต basic fields
        if (reportDto.getReportTitle() != null) existingReport.setReportTitle(reportDto.getReportTitle());
        if (reportDto.getReportMessage() != null) existingReport.setReportMessage(reportDto.getReportMessage());
        if (reportDto.getReportDate() != null) existingReport.setReportDate(reportDto.getReportDate());
        if (reportDto.getReportStatus() != null) existingReport.setReportStatus(reportDto.getReportStatus());

        // 2. อัปเดตความสัมพันธ์
        Integer newReporterId = reportDto.getReporter() != null ? reportDto.getReporter().getId() : null;
        Integer newPenaltyId = reportDto.getPenalty() != null ? reportDto.getPenalty().getPenaltyId() : null;

        // อัปเดต Hire (ถ้ามีการส่ง HireDTO มา)
        if (reportDto.getHire() != null && reportDto.getHire().getHireId() != null) {
            Hire newHire = hireRepository.findById(reportDto.getHire().getHireId())
                    .orElseThrow(() -> new IllegalArgumentException("Hire not found with ID: " + reportDto.getHire().getHireId()));
            existingReport.setHire(newHire);
        }

        if (newReporterId != null) {
            PartyRole newReporter = partyRoleRepository.findById(newReporterId)
                    .orElseThrow(() -> new IllegalArgumentException("Reporter not found with ID: " + newReporterId));
            existingReport.setReporter(newReporter);
        }

        // ❌ ลบ: ลอจิกการอัปเดต Hirer/Housekeeper โดยตรงออก

        if (newPenaltyId != null) {
            Penalty newPenalty = penaltyRepository.findById(newPenaltyId)
                    .orElseThrow(() -> new IllegalArgumentException("Penalty not found with ID: " + newPenaltyId));
            existingReport.setPenalty(newPenalty);
        } else if (reportDto.getPenalty() == null) {
            existingReport.setPenalty(null);
        }

        Report savedReport = reportRepository.save(existingReport);
        initializeReport(savedReport);
        return reportMapper.toDto(savedReport);
    }

    @Override
    @Transactional
    public void deleteReport(int id) {
        reportRepository.deleteById(id);
    }

    @Override
    public List<ReportDTO> getReportsByStatus(String reportStatus) {
        List<Report> reports = reportRepository.findByReportStatus(reportStatus);
        reports.forEach(this::initializeReport);
        return reportMapper.toDtoList(reports);
    }

    @Override
    public Optional<ReportDTO> findByPenaltyId(Integer penaltyId) {
        if (penaltyId == null) {
            return Optional.empty();
        }
        Optional<Report> reportOptional = reportRepository.findByPenalty_PenaltyId(penaltyId);
        reportOptional.ifPresent(this::initializeReport);

        return reportOptional.map(reportMapper::toDto);
    }

    @Override
    public Optional<ReportDTO> findLatestReportWithPenaltyByPersonId(Integer personId) {
        // [Note: Assuming findReportsWithPenaltyByPersonId is a repository method]
        List<Report> reports = reportRepository.findReportsWithPenaltyByPersonId(personId);
        if (!reports.isEmpty()) {
            Report latestReport = reports.get(0);
            initializeReport(latestReport);
            return Optional.of(reportMapper.toDto(latestReport));
        }
        return Optional.empty();
    }

    @Override
    public Optional<ReportDTO> findByHireIdAndReporterId(Integer hireId, Integer reporterId) {
        if (hireId == null || reporterId == null) {
            return Optional.empty();
        }
        Optional<Report> reportOptional = reportRepository.findByHire_HireIdAndReporter_Id(hireId, reporterId);
        reportOptional.ifPresent(this::initializeReport);
        return reportOptional.map(reportMapper::toDto);
    }

    @Override
    @Transactional
    public void updateUserAccountStatus(int personId, boolean isBanned) {
        String newStatus = isBanned ? "banned" : "active";
        personService.updateAccountStatus(personId, newStatus);
    }
}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/report/service/impl/ReportServiceImpl.kt
package com.itsci.mju.maebanjumpen.report.service.impl

import com.itsci.mju.maebanjumpen.entity.*
import com.itsci.mju.maebanjumpen.hire.dto.HireDTO
import com.itsci.mju.maebanjumpen.hire.repository.HireRepository
import com.itsci.mju.maebanjumpen.hire.service.impl.HireStatusUpdateService
import com.itsci.mju.maebanjumpen.partyrole.dto.*
import com.itsci.mju.maebanjumpen.partyrole.repository.PartyRoleRepository
import com.itsci.mju.maebanjumpen.penalty.dto.PenaltyDTO
import com.itsci.mju.maebanjumpen.penalty.repository.PenaltyRepository
import com.itsci.mju.maebanjumpen.person.service.PersonService
import com.itsci.mju.maebanjumpen.report.dto.ReportDTO
import com.itsci.mju.maebanjumpen.report.repository.ReportRepository
import com.itsci.mju.maebanjumpen.report.request.CreateReportRequest
import com.itsci.mju.maebanjumpen.report.request.UpdateReportRequest
import com.itsci.mju.maebanjumpen.report.service.ReportService
import com.luca.intern.common.exception.NotFoundException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.Optional

@Service
@Transactional(readOnly = true)
class ReportServiceImpl @Autowired internal constructor(
  private val reportRepository: ReportRepository,
  private val partyRoleRepository: PartyRoleRepository,
  private val penaltyRepository: PenaltyRepository,
  private val hireRepository: HireRepository,
  private val personService: PersonService,
  private val hireStatusUpdateService: HireStatusUpdateService
) : ReportService {

  private fun mapPartyRoleToDto(partyRole: PartyRole): PartyRoleDTO {
    return when (partyRole) {
      is Hirer -> HirerDTO().apply {
        id = partyRole.id?.toInt()
        balance = partyRole.balance
      }
      is Housekeeper -> HousekeeperDTO().apply {
        id = partyRole.id?.toInt()
        balance = partyRole.balance
        photoVerifyUrl = partyRole.photoVerifyUrl
        statusVerify = partyRole.statusVerify?.name
        rating = partyRole.rating
        dailyRate = partyRole.dailyRate
      }
      is Admin -> AdminDTO().apply {
        id = partyRole.id?.toInt()
        adminStatus = partyRole.adminStatus
      }
      is AccountManager -> AccountManagerDTO().apply {
        id = partyRole.id?.toInt()
        managerID = partyRole.managerID?.toInt()
      }
      is Member -> MemberDTO().apply {
        id = partyRole.id?.toInt()
        balance = partyRole.balance
      }
      else -> MemberDTO().apply {
        id = partyRole.id?.toInt()
      }
    }
  }

  private fun mapReportToDto(report: Report): ReportDTO {
    return ReportDTO(
        id = report.id,
        reportTitle = report.reportTitle,
        reportMessage = report.reportMessage,
        reportDate = report.reportDate,
        reportStatus = report.reportStatus,
        reporter = report.reporter?.let { mapPartyRoleToDto(it) },
        penalty = report.penalty?.let {
          PenaltyDTO(
              id = it.id,
              penaltyType = it.penaltyType,
              penaltyDetail = it.penaltyDetail,
              penaltyDate = it.penaltyDate,
              penaltyStatus = it.penaltyStatus
          )
        },
        hire = report.hire?.let { HireDTO(id = it.id, hireName = it.hireName, jobStatus = it.jobStatus) }
    )
  }

  override fun listAllReports(): List<ReportDTO> {
    return reportRepository.findAll().map { mapReportToDto(it) }
  }

  @Transactional
  override fun createReport(request: CreateReportRequest): ReportDTO {
    partyRoleRepository.findById(request.reporterId)
        .orElseThrow { IllegalArgumentException("Reporter not found with id: ${request.reporterId}") }
    hireRepository.findById(request.hireId)
        .orElseThrow { IllegalArgumentException("Hire not found with id: ${request.hireId}") }

    val report = Report(
        reportTitle = request.reportTitle ?: "",
        reportMessage = request.reportMessage ?: "",
        reportDate = LocalDateTime.now(),
        reportStatus = request.reportStatus ?: "",
        reporterId = request.reporterId,
        hireId = request.hireId
    )
    val savedReport = reportRepository.save(report)
    return mapReportToDto(savedReport)
  }

  override fun getReportById(reportId: Long): ReportDTO {
    val report = reportRepository.findById(reportId)
        .orElseThrow { IllegalArgumentException("Report not found with id: $reportId") }
    return mapReportToDto(report)
  }

  @Transactional
  override fun deleteReport(id: Long): Boolean {
    reportRepository.findById(id)
        .orElseThrow { IllegalArgumentException("Report not found with id: $id") }
    reportRepository.deleteById(id)
    return true
  }

  override fun getReportsByStatus(reportStatus: String): List<ReportDTO> {
    return reportRepository.findByReportStatus(reportStatus).map { mapReportToDto(it) }
  }

  @Transactional
  override fun updateReport(request: UpdateReportRequest): ReportDTO {
    val report = reportRepository.findById(request.id)
        .orElseThrow { IllegalArgumentException("Report not found with id: ${request.id}") }

    request.reportTitle?.let { report.reportTitle = it }
    request.reportMessage?.let { report.reportMessage = it }
    request.reportStatus?.let { report.reportStatus = it }
    if (request.reporterId > 0) {
      partyRoleRepository.findById(request.reporterId)
          .orElseThrow { IllegalArgumentException("Reporter not found with id: ${request.reporterId}") }
      report.reporterId = request.reporterId
    }

    val savedReport = reportRepository.save(report)
    return mapReportToDto(savedReport)
  }

  override fun findByPenaltyId(penaltyId: Long?): Optional<ReportDTO> {
    if (penaltyId == null) return Optional.empty()
    return reportRepository.findByPenalty_PenaltyId(penaltyId)
        .map { mapReportToDto(it) }
  }

  override fun findLatestReportWithPenaltyByPersonId(personId: Long?): Optional<ReportDTO> {
    if (personId == null) return Optional.empty()
    val reports = reportRepository.findReportsWithPenaltyByPersonId(personId)
    return if (reports.isNotEmpty()) {
      Optional.of(mapReportToDto(reports.first()))
    } else {
      Optional.empty()
    }
  }

  @Transactional
  override fun updateUserAccountStatus(personId: Long, isBanned: Boolean) {
    val person = personService.getPersonById(personId)
        ?: throw IllegalArgumentException("Person not found with id: $personId")
    person.accountStatus = if (isBanned) "banned" else "active"
    personService.updatePerson(personId, person)
  }


}

